toml/src/main/kotlin/org/rust/toml/inspections/CargoTomlCyclicFeatureInspection.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.inspections
+
+import com.intellij.codeInspection.LocalInspectionTool
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElementVisitor
+import org.rust.cargo.CargoConstants
+import org.rust.toml.isFeatureListHeader
+import org.toml.lang.psi.TomlArray
+import org.toml.lang.psi.TomlLiteral
+import org.toml.lang.psi.TomlTable
+import org.toml.lang.psi.TomlVisitor
+import org.toml.lang.psi.ext.TomlLiteralKind
+import org.toml.lang.psi.ext.kind
+
+/**
+ *  * Consider `Cargo.toml`:
+ * ```
+ * [features]
+ * foo = ["foo"]
+ *       #^ Shows error that "foo" feature depends on itself
+ * ```
+ */
+class CargoTomlCyclicFeatureInspection : LocalInspectionTool() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
+        buildVisitor(holder) ?: super.buildVisitor(holder, isOnTheFly)
+
+    private fun buildVisitor(holder: ProblemsHolder): PsiElementVisitor? {
+        if (holder.file.name != CargoConstants.MANIFEST_FILE) return null
+
+        return object : TomlVisitor() {
+            override fun visitTable(element: TomlTable) {
+                if (!element.header.isFeatureListHeader) return
+
+                for (keyValue in element.entries) {
+                    val parentFeatureName = keyValue.key.text ?: continue
+
+                    val featureArray = keyValue.value as? TomlArray ?: continue
+                    for (feature in featureArray.elements) {
+                        val featureName = ((feature as? TomlLiteral)?.kind as? TomlLiteralKind.String)?.value
```

Comment:
In IntelliJ it's always better to attach annotations directly to the `element` that is visited now. it's a bit long story... And we often don't adhere to this rule. But let's try it in this case since there is a visitor and it looks pretty easy. Let's override `visitLiteral` and then check that it is a string in TomlArray in TomlKeyValue in TomlTable with a header that is `isFeatureListHeader`.